source/assets/js/playground/autocomplete.ts

```diff
@@ -0,0 +1,245 @@
+  'error',
+  'warn',
+  'debug',
+  'at-root',
+  'if',
+  'else',
+  'each',
+  'for',
+  'while',
+];
+
+// CompletionResult options for Sass at rules, for example `@use`.
+const atRuleOptions = Object.freeze(
+  atRuleKeywords.map(keyword => ({
+    label: `@${keyword} `,
+    type: 'keyword',
+    validFor: identifier,
+  }))
+);
+
+// Completions for Sass at rules
+function atRuleCompletion(context: CompletionContext): CompletionResult | null {
+  const atRule = context.matchBefore(/@\w*/);
+  if (!atRule) return null;
+  if (atRule.from === atRule.to && !context.explicit) return null;
+  return {
+    from: atRule.from,
+    to: atRule.to,
+    options: atRuleOptions,
+    validFor: identifier,
+  };
+}
+
+// A list of all the Sass built in modules.
+const moduleNames = moduleMetadata.map(mod => mod.name);
+type ModuleName = (typeof moduleNames)[number];
+
+// Matches any of the built in Sass modules.
+const moduleNameRegExp = new RegExp(`(${moduleNames.join('|')}).\\$?\\w*`);
+
+// Matches the StringLiteral `"sass:modName"`, capturing `modName`.
+const moduleUseRegex = new RegExp(/['"]sass:(?<modName>.*)['"]/);
+
+// Completion results for built in Sass modules following the `sass:` namespace.
+const moduleCompletions = Object.freeze(
+  moduleMetadata.map(mod => ({
+    label: `sass:${mod.name}`,
+    apply: `sass:${mod.name}`,
+    info: mod.description,
+    type: 'class',
+    validFor: identifier,
+    _moduleName: mod.name as ModuleName,
+  }))
+);
+
+// Completions for the import of built in modules, for instance "sass:color".
+function moduleMetadataCompletion(
+  context: CompletionContext
+): CompletionResult | null {
+  const nodeBefore = syntaxTree(context.state).resolveInner(context.pos, -1);
+  const potentialTypes = [
+    'StringLiteral', // When wrapped in quotes- `"sass:"`
+    'ValueName', // No end quote, before semicolon- `"sa`
+    ':', // No end quote, on semicolon- `"sass:`
+    'PseudoClassName', // No end quote, after semicolon- `"sass:m`
+  ];
+  if (!potentialTypes.includes(nodeBefore.type.name)) return null;
+  const potentialParentTypes = [
+    'UseStatement', // When wrapped in quotes
+    'PseudoClassSelector', // No end quote, after the colon
+  ];
+  if (!potentialParentTypes.includes(nodeBefore.parent?.type.name || '')) {
+    return null;
+  }
+
+  const moduleMatch = context.matchBefore(/["'](sass:)?\w*/);
+
+  if (!moduleMatch) return null;
+  if (moduleMatch.from === moduleMatch.to && !context.explicit) return null;
+
+  const included = includedmoduleMetadata(context.state);
+  const notIncludedModuleCompletions = moduleCompletions.filter(
+    moduleCompletion => !included.includes(moduleCompletion._moduleName)
+  );
+  return {
+    from: moduleMatch.from + 1,
+    to: moduleMatch.to,
+    options: notIncludedModuleCompletions,
+    validFor: identifier,
+  };
+}
+
+// Completion results for module variables.
+const moduleVariableCompletions = Object.freeze(
+  moduleMetadata.reduce(
```

Comment:
If we're going to keep the modules as a list of objects, I'd make this clearer by creating a helper function like `mapModulesByName<V>(value: (module: moduleMetadata[number]) => V): Record<K, V>`

Reply:
Added in https://github.com/sass/sass-site/pull/1166/commits/ee7ac67c8677fd8b6bae2e863ff3086f55264165.